### PR TITLE
Add support for Nested Conditions via AndCondition and OrCondition 

### PR DESCRIPTION
**Pull request by Toflar**
Started implementing https://github.com/schranz-search/schranz-search/issues/317.

Works for Meilisearch and Loupe, the others I haven't got any local setup for at the moment 🙈 
But maybe this can be a helpful start to push it over the finishing line.

---

This adds 2 new conditions.

**AndCondition**

```php
$searchBuilder->addFilter(new Condition\AndCondition($conditionA, $conditionB);
```

**OrCondition**

```php
$searchBuilder->addFilter(new Condition\OrCondition($conditionA, $conditionB);
```

----

**Nested conditions** are now possible:

```php
$searchBuilder->addFilter(
    new Condition\AndCondition(
        new Condition\EqualCondition('tags', 'Tech'),
        new Condition\OrCondition(
            new Condition\EqualCondition('tags', 'UX'),
            new Condition\EqualCondition('isSpecial', false),
        ),
    ),
);
```

### CODE REVIEW

**Comment by alexander-schranz** [comment]
@toflar Thank you!

> Works for Meilisearch and Loupe, the others I haven't got any local setup for at the moment 🙈

If you want to give it a try a `docker-compose.yml` exists in every adapters directory if you want to start the service `docker compose up` is enough. What may very similar to Meilisearch is Typesense Adapter: https://github.com/schranz-search/schranz-search/blob/4267ea93ec459e15df084cd1727457dcbd1a4457/packages/seal-typesense-adapter/src/TypesenseSearcher.php#L77-L114

The only problem if I see currently a Geo based Conditions can not be part of a Nested Conditions. But think that is some edgecase we currently can ignore.

File: packages/seal/src/Search/Condition/AndCondition.php [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace Schranz\Search\SEAL\Search\Condition;
+
+class AndCondition
+{
+    /**
+     * @var array<object>
+     */
+    public function __construct(public readonly array $conditions)
```
Comment (alexander-schranz) [inline packages/seal/src/Search/Condition/AndCondition.php]: What you think is more developer friendly go here with an `array` or variable parameters length e.g.:

```php
public function __construct(...$conditions)
```

```php
$condition = new Condition\OrCondition(
    new Condition\AndCondition(
        new Condition\EqualCondition('tags', 'Tech'),
        new Condition\EqualCondition('isSpecial', true),
    ),
    new Condition\AndCondition(
        new Condition\EqualCondition('tags', 'UX'),
        new Condition\EqualCondition('isSpecial', false),
    ),
);
```

**Comment by Toflar** [comment]
I have updated the conditions so that they are easier to use and they also only accept the supported nested conditions. Typesense is implemented as well now.

For the `MemoryAdapter` - am I right that this is just for testing purposes? We could just use Loupe with the in-memory configuration - it's been designed for exactly that and then we wouldn't have to develop all that logic with nested comparisons etc.

**Comment by alexander-schranz** [comment]
I will take care of the memory adapter. At current state I want to keep the memory adapter as it good to have one which don't require any additional dependencies.

**Comment by Toflar** [comment]
Elasticsearch, Opensearch and Algolia are now also supported. I had to adjust the example because Algolia does not support filtering for ORs of ANDs (https://github.com/algolia/algoliasearch-client-php/issues/385).

**Comment by alexander-schranz** [comment]
@Toflar Good work, thank you 👍

**Pull request merged**